### ISSUE TITLE
setter.variable.type is the getter's return type, not the setter's parameter type

### ISSUE THREAD

**Issue by MichaelRFairhurst** [dart-lang/sdk#29392]
If I make a class with
```
set foo(Type1 blah) => ...
get Type2 foo => null;
```

I would expect `lookupSetter("foo").variable.type` to be `Type1`. But its `Type2`?

**Comment by bwilkerson** [comment]
I think the real bug is that we shouldn't be modeling getters and setters as if they induced a variable.

But given that we are, and given that the getter's type is allowed to be more general than the setter's parameter type [1], I think it makes sense for the variable to have the getter's type.

[1] Section 10.3 of the spec reads:

> It is a static warning if a class has a setter named v = with argument type T and a getter named v with return type S, and T may not be assigned to S.

**Comment by MichaelRFairhurst** [comment]
Hmm, interesting in terms of T being assignable to S.

However, not all values assignable to S must be assignable to T.

The material components library has a lot of cases like this:

```
  bool _flag;
  set flag(dynamic flag) => _flag = coerceBoolean(flag);
  bool get flag => _flag;
```

This led to errors where `flag` was assigned a `String` (which is the purpose of `coerceBoolean`), which is of course legal dart, but my implementation resulted in a type error.

```
  // assuming we coded a typecheck up as
  if (!type.isAssignableTo(setter.variable.type)) {
    throw ...
  }

  // shouldn't this become equivalent to
  if (StringType.isAssignableTo(dynamicType)) {
    // will pass, no exception thrown, matches assignability rules
  }

  // instead of
  if (StringType.isAssignableTo(boolType)) {
    // won't pass, not matching assignability rules
  }
```

But it does get to be a bit of a weird question in the first place given that the variable is induced only.

**Comment by bwilkerson** [comment]
Yeah, it's a bit weird. I'd recommend that you only use the induced variable as an easy way to find the corresponding getter.

Two other items of note. First, the rules about what is assignable to what are different under strong mode. Second, I'm not sure that `DartType.isAssignableTo` has been updated to work correctly in strong mode; it might still be using non-strong mode rules. To ensure that you have the right semantics you need to get access to the `TypeSystem` associated with the context.

**Comment by bwilkerson** [comment]
One other thought. When analyzer is analyzing code like
```
foo.bar = baz;
```
It notes that the setter `bar` is being invoked and uses the parameter type to compare with the type of the right-hand side. But when it's analyzing code like
```
f(foo.bar);
```
It notes that the getter `bar` is being invoked and uses the return type. That allows mismatches between the two to be correctly checked based on the context.

**Comment by MichaelRFairhurst** [comment]
Hmm, we're using this API to take:

```
class foo {
  @Input()
  String b;
}
```

and typecheck

```
<foo [b]="bar">
```

Previously we were using `.variable`, but given this I changed it to `member.type.parameters[0]`. Seems uglier to read, but it works.
